flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java

```diff
@@ -0,0 +1,187 @@
+            } else {
+                return FutureUtils.completedExceptionally(
+                        new IllegalStateException(
+                                "Task "
+                                        + executionsWithGateway.f0.getTaskNameWithSubtaskIndex()
+                                        + " is not running."));
+            }
+        }
+
+        synchronized (lock) {
+            if (isShutDown) {
+                return FutureUtils.completedExceptionally(new IllegalStateException("Shut down"));
+            }
+
+            final int requestId = requestIdCounter++;
+
+            log.debug("Triggering thread info request {}", requestId);
+
+            final PendingThreadInfoRequest pending =
+                    new PendingThreadInfoRequest(requestId, runningSubtasksIds);
+
+            // requestTimeout is treated as the time on top of the expected sampling duration.
+            // Discard the request if it takes too long. We don't send cancel
+            // messages to the task managers, but only wait for the responses
+            // and then ignore them.
+            long expectedDuration = numSamples * delayBetweenSamples.toMilliseconds();
+            Time timeout = Time.milliseconds(expectedDuration + requestTimeout.toMilliseconds());
+
+            // Add the pending request before scheduling the discard task to
+            // prevent races with removing it again.
+            pendingRequests.put(requestId, pending);
+
+            ThreadInfoSamplesRequest requestParams =
+                    new ThreadInfoSamplesRequest(
+                            requestId, numSamples, delayBetweenSamples, maxStackTraceDepth);
+
+            requestThreadInfo(subtasksWithGateways, requestParams, timeout);
+
+            return pending.getStatsFuture();
+        }
+    }
+
+    /**
+     * Requests thread infos from given subtasks. The response would be ignored if it does not
+     * return within timeout.
+     */
+    private void requestThreadInfo(
+            List<Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>>
+                    subtasksWithGateways,
+            ThreadInfoSamplesRequest requestParams,
+            Time timeout) {
+
+        // Trigger samples collection from all subtasks
+        for (Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>
+                executionWithGateway : subtasksWithGateways) {
+
+            CompletableFuture<TaskExecutorGateway> executorGatewayFuture = executionWithGateway.f1;
+
+            ExecutionAttemptID taskExecutionAttemptId =
+                    executionWithGateway.f0.getCurrentExecutionAttempt().getAttemptId();
+
+            CompletableFuture<TaskThreadInfoResponse> threadInfo =
+                    executorGatewayFuture.thenCompose(
+                            executorGateway ->
+                                    executorGateway.requestThreadInfoSamples(
```

Comment:
this will send an RPC for every attempt. To reduce the number of issues RPC calls we could batch requests for executions running in the same task executor.

Reply:
Good point. Unfortunately I won't be able to address this optimization before the feature freeze because I am currently on vacation. How about we add it as an "improvement" ticket.

Reply:
👍  for follow-up work.